crates/ruff_linter/src/rules/flake8_trio/rules/zero_sleep_call.rs

```diff
@@ -0,0 +1,109 @@
+use ruff_diagnostics::{AlwaysFixableViolation, Diagnostic, Edit, Fix};
+use ruff_macros::{derive_message_formats, violation};
+use ruff_python_ast::Stmt;
+use ruff_python_ast::{self as ast, Expr, ExprCall, Int};
+use ruff_text_size::Ranged;
+
+use crate::checkers::ast::Checker;
+use crate::importer::ImportRequest;
+
+/// ## What it does
+/// Checks for calls of `trio.sleep(0)` and suggests replacement with `trio.lowlevel.checkpoint()`
+///
+/// ## Why is this bad?
+/// Although equivalent to `trio.sleep(0)`, a call to `trio.lowlevel.checkpoint()` is more
+/// suggestive of the underlying process and the author's intent.
+///
+/// ## Example
+/// ```python
+/// async def func():
+///     await trio.sleep(0)
+/// ```
+///
+/// Use instead:
+/// ```python
+/// async def func():
+///     await trio.lowlevel.checkpoint()
+/// ```
+#[violation]
+pub struct TrioZeroSleepCall;
+
+impl AlwaysFixableViolation for TrioZeroSleepCall {
+    #[derive_message_formats]
+    fn message(&self) -> String {
+        format!("Use `trio.lowlevel.checkpoint()` instead of `trio.sleep(0)`")
+    }
+
+    fn fix_title(&self) -> String {
+        format!("Replace `trio.sleep(0)` with `trio.lowlevel.checkpoint()`")
+    }
+}
+
+/// TRIO115
+pub(crate) fn zero_sleep_call(checker: &mut Checker, call: &ExprCall) {
+    if !checker
+        .semantic()
+        .resolve_call_path(call.func.as_ref())
+        .is_some_and(|call_path| matches!(call_path.as_slice(), ["trio", "sleep"]))
+    {
+        return;
+    }
+
+    if call.arguments.len() != 1 {
+        return;
+    }
+
+    let Some(arg) = call.arguments.find_argument("seconds", 0) else {
+        return;
+    };
+
+    match arg {
+        Expr::NumberLiteral(ast::ExprNumberLiteral { value, .. }) => {
+            let Some(int) = value.as_int() else { return };
+            if *int != Int::ZERO {
+                return;
+            }
+        }
+        Expr::Name(ast::ExprName { id, .. }) => {
+            let scope = checker.semantic().current_scope();
```

Comment:
I couldn't find a more compact way to do this yet. In case there is one somewhere in the repo already, happy to hear about it and refactor this.